drivers/led_strip/ws2812_uart.c

```diff
@@ -0,0 +1,232 @@
+/* which will be used to send ws2812 rgb led */
+static const uint8_t ws2812_uart_lut[8] = {
+	0b1011011,
+	0b0011011,
+	0b1010011,
+	0b0010011,
+	0b1011010,
+	0b0011010,
+	0b1010010,
+	0b0010010,
+};
+
+/*
+ * Serialize an 24-bit color channel value into an equivalent sequence
+ * of 8 bytes of uart frames,.
+ */
+static inline void ws2812_uart_ser(uint8_t *buf, uint8_t color[3])
+{
+	int i, j;
+	int buff_indx = 0;
+	uint8_t serialized_val = 0;
+
+	for (i = 0; i < 3; i++) {
+		for (j = 0; j < 8; j++) {
+			serialized_val = (serialized_val << 1) | ((color[i]>>(7-j))&0x01);
+			if (((i*8)+j) % 3 == 2) {
+				buf[buff_indx++] = ws2812_uart_lut[serialized_val];
+				serialized_val = 0;
+			}
+		}
+	}
+}
+
+/*
+ * Returns true if and only if cfg->px_buf is big enough to convert
+ * num_pixels RGB color values into UART frames.
+ */
+static inline bool num_pixels_ok(const struct ws2812_uart_cfg *cfg,
+				 size_t num_pixels)
+{
+	size_t nbytes;
+	bool overflow;
+
+	overflow = size_mul_overflow(num_pixels, cfg->num_colors * 8, &nbytes);
+	return !overflow && (nbytes <= cfg->px_buf_size);
+}
+
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
+
+static void ws2812_uart_tx(const struct ws2812_uart_cfg *cfg, uint8_t *tx, size_t len)
+{
+	for (size_t i = 0; i < len; i++) {
+		uart_poll_out(cfg->uart_dev, tx[i]);
+	}
+}
+static int ws2812_strip_update_rgb(const struct device *dev,
+				   struct led_rgb *pixels,
+				   size_t num_pixels)
+{
+	const struct ws2812_uart_cfg *cfg = dev->config;
+
+	uint8_t *px_buf = cfg->px_buf;
+	size_t px_buf_len = cfg->px_buf_size;
+	size_t uart_data_len = 0;
+	size_t i;
+	int rc = 0;
+	uint8_t color[3];
+
+	if (!num_pixels_ok(cfg, num_pixels)) {
+		return -ENOMEM;
+	}
+
+	/*
+	 * Convert pixel data into Uart frames. Each frame has pixel data
+	 * in color mapping on-wire format (e.g. GRB, GRBW, RGB, etc).
+	 */
+
+	for (i = 0; i < num_pixels; i++) {
+		uint8_t j;
+
+		for (j = 0; j < cfg->num_colors; j++) {
+			switch (cfg->color_mapping[j]) {
+			/* White channel is not supported by LED strip API. */
+			case LED_COLOR_ID_WHITE:
+				color[j] = 0;
+				break;
+			case LED_COLOR_ID_RED:
+				color[j] = pixels[i].r;
+				break;
+			case LED_COLOR_ID_GREEN:
+				color[j] = pixels[i].g;
+				break;
+			case LED_COLOR_ID_BLUE:
+				color[j] = pixels[i].b;
+				break;
+			default:
+				return -EINVAL;
+			}
+		}
+		/*Convert 24 bits of rgb data into 8 bytes of uart frame*/
+		ws2812_uart_ser(px_buf, &color[0]);
+		uart_data_len += 8;
+		px_buf += uart_data_len;
+	}
+	ws2812_uart_tx(cfg, cfg->px_buf, uart_data_len);
+	ws2812_reset_delay(51);
+
+	return rc;
+}
+
+static int ws2812_strip_update_channels(const struct device *dev,
+					uint8_t *channels,
+					size_t num_channels)
+{
+	LOG_ERR("update_channels not implemented");
+	return -ENOTSUP;
+}
```

Comment:
function can go

Reply:
I dont think we should keep this api NULL because the user may use this API without checking if its NULL or not.
Also, doing so will let user know that this feature is not supported at the moment with the `error` return code `-ENOTSUP`.